tests/test_upgrade.py

```diff
@@ -0,0 +1,608 @@
+def test_minting_disabled_but_preupgrade_beth_are_withdrawable(
+    accounts,
+    stranger,
+    steth_token,
+    lido_oracle_report,
+    deploy_vault_and_pass_dao_vote,
+    deposit_amount,
+    steth_approx_equal,
+):
+    """
+    This is a big integrated test that validates,
+        a) minting after the v4 upgrade is disabled,
+        b) any bETH that were minted before the upgrade are still withdrawable,
+        c) the 2022_01_26 incident mitigation is not performed again.
+
+    Why is this test so big?
+
+        Rather than dividing this test into smaller unit tests,
+        we believe, that a big integrated test will better illustrate
+        what the v4 upgrade entails and will show a more cohesive picture.
+
+    This test runs in several stages.
+
+        Stage 0. Setup - declare all the necessary variables and initialize contracts.
+
+        Stage 1. Pre-upgrade - validate what we will change
+        in this upcoming upgrade. This will help us understand
+        what has changed after the upgrade.
+
+        Stage 2. Upgrade - validate post-upgrade configuration
+
+        Stage 3. 2022/01/26 - confirm that the mitigation upgrade (version 3)
+        for the 2022/01/26 burned bETH incident has not been performed again.
+
+        Stage 4. Mint - confirm that minting doesn't work. `submit` reverts
+        and both beth and steth total supplies remain the same.
+
+        Stage 5. Rebase - simulate positive stETH rebase. The withdrawal rate only changes
+        in case of a negative rebase, but we will still simulate a positive rebase
+        to better simulate an actual behavior of the protocol.
+
+        Stage 5. Withdraw - confirm that the user can withdraw their bETH worth of stETH
+        that were minted on Terra before the upgrade. We will simulate the bridging of funds
+        from Wormhole and call `withdraw` from stranger.`
+
+    """
+
+    ##################
+    # STAGE 0. Setup #
+    ##################
+
+    # initialize vault proxy
+    vault_proxy = brownie.Contract.from_abi(
+        "AnchorVaultProxy", config.vault_proxy_addr, brownie.AnchorVaultProxy.abi
+    )
+
+    # initialize vault
+    vault = brownie.Contract.from_abi(
+        "AnchorVault", config.vault_proxy_addr, brownie.AnchorVault.abi
+    )
+
+    beth_token = brownie.interface.ERC20(vault.beth_token())
+
+    liquidations_admin = accounts.at(vault.liquidations_admin(), force=True)
+
+    ########################
+    # STAGE 1. Pre-upgrade #
+    ########################
+
+    # minting new bETH works as expected
+
+    prev_beth_total_supply = beth_token.totalSupply()
+
+    vault.submit(
+        deposit_amount,
+        TERRA_ADDRESS,
+        "0x8bada2e",
+        vault.version(),
+        {"from": stranger, "value": deposit_amount},
+    )
+
+    # before and after delta is stranger's balance, as we can't directly query stranger's balance
+    preupgrade_terra_beth_minted_to_stranger = (
```

Comment:
minor: would be more obvious to get it from the `Deposited` event's `beth_amount_received` field